Add metrics and alerts tutorial to the docs (#6341)
operator-framework/operator-sdk: website/content/en/docs/building-operators/golang/advanced-topics.md

* After adding new import paths to your operator project, run `go mod vendor` if a `vendor/` directory is present in the root of your project directory to fulfill these dependencies.
* Your 3rd party resource needs to be added before add the controller in `"Setup all Controllers"`.

### Monitoring and Observability
This section aims to explain how to expose the first operator's custom metrics and alerts. It focuses on the technical aspects of adding metrics and alerts, and demonstrates it over the sample [memcached-operator]. For more information regarding operators' monitoring, visit [observability-best-practices].  

#### Prerequisites

> - Deploy [kube-prometheus] on your cluster.

Author: It's possible to export the metrics and view them using the metrics endpoint, without Prometheus installed.
But AFAIK, Prometheus must be installed in order to check the alerts and recording rules, so I think it should be a prerequisite for this tutorial. 
I can add an explanation about the metrics endpoint in addition.

> - Deploy [kube-prometheus] on your cluster.

Reviewer: So think you should comment something like: 
These pre-requisites are needed in order to:
1. Save metrics to Prometheus
2. Handle Prometheus alerts  based on the alerting rules [Link to Prometheus docs].
3. Calculate Prometheus recording rules [link to Prometheus docs]

> - Deploy [kube-prometheus] on your cluster.

Author: updated

- Make sure Prometheus has access to the operator's namespace, by setting the corresponding RBAC rules. You may find [prometheus_role.yaml] and [prometheus_role_binding.yaml] examples helpful. 

> - Create a new `monitoring` directory, which will contain the metrics and alerts implementation.

Reviewer: Please explain why.
For example: We recommend to create a new `monitoring` directory, which will contain the metrics and alerts implementation, in order to keep the monitoring code and logic separated from the core operator code. This will allow easier maintenance of both the core operator code and the monitoring code.

Author: updated

- Add the following command to the `dockerfile`:
 ```dockerfile
COPY monitoring/ monitoring/
 ```

#### Publishing Custom Metrics
If you wish to publish custom metrics for your operator, this can be easily achieved by using the global registry from `controller-runtime/pkg/metrics`.
One way to achieve this is to declare your collectors as global variables, register them using `RegisterMetrics()` and call it in the controller's `init()` function.

Example custom metric: [MemcachedDeploymentSizeUndesiredCountTotal].

```go
package monitoring

import (
	"github.com/prometheus/client_golang/prometheus"
	"sigs.k8s.io/controller-runtime/pkg/metrics"
)

var (
	MemcachedDeploymentSizeUndesiredCountTotal = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "memcached_deployment_size_undesired_count_total",
			Help: "Total number of times the deployment size was not as desired.",
		},
	)
)

// RegisterMetrics will register metrics with the global prometheus registry
func RegisterMetrics() {
	metrics.Registry.MustRegister(MemcachedDeploymentSizeUndesiredCountTotal)
}
```

> 

Reviewer: Please add here to also go to the best practices document to see the metrics best practices, For example for naming conventions.

Author: Done

[PrometheusRule reconciliation example] example:
```go
func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
    ...
    ...
    // Check if prometheus rule already exists, if not create a new one
    foundRule := &monitoringv1.PrometheusRule{}
    err := r.Get(ctx, types.NamespacedName{Name: ruleName, Namespace: namespace}, foundRule)
    if err != nil && apierrors.IsNotFound(err) {
        // Define a new prometheus rule
        prometheusRule := monitoring.NewPrometheusRule(namespace)
        if err := r.Create(ctx, prometheusRule); err != nil {
            log.Error(err, "Failed to create prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    
    if err == nil {
    // Check if prometheus rule spec was changed, if so set as desired
    desiredRuleSpec := monitoring.NewPrometheusRuleSpec()
    if !reflect.DeepEqual(foundRule.Spec.DeepCopy(), desiredRuleSpec) {
        desiredRuleSpec.DeepCopyInto(&foundRule.Spec)
        if r.Update(ctx, foundRule); err != nil {
            log.Error(err, "Failed to update prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    ...
    ...
}
```

> - For more information regarding alerts annotations and labels, follow [observability-best-practices].

Reviewer: For more information regarding alerts and recording rules best practices, please follow the [observability-best-practices].

Author: Done

- It is highly recommended implementing unit tests for prometheus rules. For more information, visit Prometheus [unit testing documentation]. For a reference of unit tests implementation in a Golang operator, see memcached-operator [alerts unit tests].

> 

Reviewer: Please add here a sub header for inspecting the metrics, recording rules and alerts locally in the Prometheus UI.

Author: It's already explained on line 295. Or do you mean something else?

Reviewer: But there is no sub header

Author: Done

* After adding new import paths to your operator project, run `go mod vendor` if a `vendor/` directory is present in the root of your project directory to fulfill these dependencies.
* Your 3rd party resource needs to be added before add the controller in `"Setup all Controllers"`.

### Monitoring and Observability
This section aims to explain how to expose the first operator's custom metrics and alerts. It focuses on the technical aspects of adding metrics and alerts, and demonstrates it over the sample [memcached-operator]. For more information regarding operators' monitoring, visit [observability-best-practices].  

#### Prerequisites
- Deploy [kube-prometheus] on your cluster. This will set a Prometheus instance on the cluster, which will enable the Prometheus UI, in which we will inspect the metrics, [alerts] and [recording rules].
- Make sure Prometheus has access to the operator's namespace, by setting the corresponding RBAC rules. You may find [prometheus_role.yaml] and [prometheus_role_binding.yaml] examples helpful. 

> - Create a new `monitoring` directory, which will contain the metrics and alerts implementation. It is recommended to create a dedicated `monitoring` directory, in order to keep the monitoring implementation separated from the core operator's logic. This will allow easier maintenance of both the core operator code and the monitoring code. 

Reviewer: I think we need to refactor this a bit to be clearer..
Perhaps:

#### Prerequisites

The following are required in order to inspect the metric, [alerts] and [recording rules], in Prometheus, during the development:
- Deploy [kube-prometheus] on your cluster.
- Enable Prometheus access to the required namespace, by setting the RBAC rules.  
  Example:  [prometheus_role.yaml] and [prometheus_role_binding.yaml].
- Create a  `/monitoring` sub-folder in your operator repository.  
**Note:** It is recommended to create a dedicated `/monitoring` directory, in order to keep the monitoring code and the core operator code separated.

Author: updated the two prerequisites, based on @varshaprasad96 comments.
Regarding the third one - please see https://github.com/operator-framework/operator-sdk/pull/6341#discussion_r1129835586

- Add the following command to the `dockerfile`:
     ```dockerfile

>     COPY monitoring/ monitoring/

Reviewer: Not sure I understand where you should run the command.. 
Is this for copying the monitoring sub folder from the Memcached to the developer operator?

Reviewer: Perhaps replace with:

Copy the example `/monitoring` sub folder from the example Memcached operator[add link to the example operator] and update it to you needs.

Author: This step is to add a line to the `dockerfile`, in order for the operator to be deployed with the monitoring implementation.

     ```

#### Publishing Custom Metrics
If you wish to publish custom metrics for your operator, this can be easily achieved by using the global registry from `controller-runtime/pkg/metrics`.
One way to achieve this is to declare your collectors as global variables, register them using `RegisterMetrics()` and call it in the controller's `init()` function.

Example custom metric: [MemcachedDeploymentSizeUndesiredCountTotal].

```go
package monitoring

import (
	"github.com/prometheus/client_golang/prometheus"
	"sigs.k8s.io/controller-runtime/pkg/metrics"
)

var (
	MemcachedDeploymentSizeUndesiredCountTotal = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "memcached_deployment_size_undesired_count_total",
			Help: "Total number of times the deployment size was not as desired.",
		},
	)
)

// RegisterMetrics will register metrics with the global prometheus registry
func RegisterMetrics() {
	metrics.Registry.MustRegister(MemcachedDeploymentSizeUndesiredCountTotal)
}
```

- The above example creates a new counter metric. For other metrics' types, visit [Prometheus Documentation].

> - For more information regarding metrics naming conventions, follow [observability-best-practices].

Reviewer: Its not only naming conventions. Please rephrase so naming convention is an example

* After adding new import paths to your operator project, run `go mod vendor` if a `vendor/` directory is present in the root of your project directory to fulfill these dependencies.
* Your 3rd party resource needs to be added before add the controller in `"Setup all Controllers"`.

### Monitoring and Observability

> This section aims to explain how to expose the first operator's custom metrics, [alerts] and [recording rules]. It focuses on the technical aspects, and demonstrates the implementation over the sample [memcached-operator]. 

Reviewer 2: Nit: What is the "first operator" in this context? IMO something like:
```suggestion
This section covers how to create custom metrics, [alerts] and [recording rules] for your operator. It focuses on the technical aspects, and demonstrates the implementation by updating the [memcached-operator] created in the [tutorial]. 
```
sounds a bit better

Author: Done, thanks

> For more information regarding operators' monitoring, please follow [observability-best-practices].  

Reviewer 2: Nit:
```suggestion
For more information regarding monitoring best practices, take a look at our docs on [observability-best-practices].  
```

Author: Done, thanks

#### Prerequisites
The following steps are required in order to inspect the operator's custom metrics, alerts and recording rules:

> - Deploy [kube-prometheus] on your cluster. This will set a Prometheus instance on the cluster, with which we will inspect the metrics, alerts and recording rules.

Reviewer 3: Do we specifically need "kube-prometheus"? Can we just say a "prometheus instance" as a pre-req? Users could use prometheus-operator or even deploy a prometheus instance directly.

Reviewer 3: We could add that using cube-prometheus is recommended. Something as in here (https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus)

Author: Done, thanks

- Make sure Prometheus has access to the operator's namespace, by setting the corresponding RBAC rules.
  
  Example: [prometheus_role.yaml] and [prometheus_role_binding.yaml]

#### Publishing Custom Metrics

> If you wish to publish custom metrics for your operator, this can be easily achieved by using the global registry from `controller-runtime/pkg/metrics`.

Reviewer 3: Do we also mention that doing this will also enable the generic controller-runtime metrics to be collected that are specified here: https://book.kubebuilder.io/reference/metrics-reference.html.

I'm not up-to-date on this though, would have to check if its enabled by default.

Author: These metrics are enabled by default

    ...
    ...
    // Check if prometheus rule already exists, if not create a new one
    foundRule := &monitoringv1.PrometheusRule{}
    err := r.Get(ctx, types.NamespacedName{Name: ruleName, Namespace: namespace}, foundRule)
    if err != nil && apierrors.IsNotFound(err) {
        // Define a new prometheus rule
        prometheusRule := monitoring.NewPrometheusRule(namespace)
        if err := r.Create(ctx, prometheusRule); err != nil {
            log.Error(err, "Failed to create prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    
    if err == nil {
    // Check if prometheus rule spec was changed, if so set as desired
    desiredRuleSpec := monitoring.NewPrometheusRuleSpec()
    if !reflect.DeepEqual(foundRule.Spec.DeepCopy(), desiredRuleSpec) {
        desiredRuleSpec.DeepCopyInto(&foundRule.Spec)
        if r.Update(ctx, foundRule); err != nil {
            log.Error(err, "Failed to update prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    ...
    ...
}
```

- Please review the [observability-best-practices] for additional important information regarding alerts and recording rules.

> - It is highly recommended implementing unit tests for prometheus rules. For more information, please follow Prometheus [unit testing documentation]. For a reference of unit tests implementation in a Golang operator, see memcached-operator [alerts unit tests].

Reviewer 2: Nit:
```suggestion
- It is highly recommended to implement unit tests for prometheus rules. For more information, please follow the Prometheus [unit testing documentation]. For examples of unit testing in a Golang operator, see the sample memcached-operator [alerts unit tests].
```

Author: I believe "recommended implementing" is correct, but besides that, followed your suggestion

#### Inspecting the metrics, alerts and recording rules with Prometheus UI

Finally, in order to inspect the exposed metrics and alerts, we need to forward  port `9090`, which corresponds with the Prometheus
deployment. This can be done with the following command:
```bash

> $ kubectl -n monitoring port-forward svc/prometheus-k8s 9090

Reviewer 3: I remember the default port for prometheus is "9090", but if I'm not wrong there is still a way to change it. Could we make it generic, like:

Finally, in order to inspect the exposed metrics and alerts, we need to forward  the corresponding port where metrics are published by Prometheus (usually "9090", which is the default value). This can be done with the following command:

```bash
$ kubectl -n monitoring port-forward svc/prometheus-k8s 9090
```

where we assume that the prometheus service is available in the "monitoring" namespace.

Author: done, thanks

        desiredRuleSpec.DeepCopyInto(&foundRule.Spec)
        if r.Update(ctx, foundRule); err != nil {
            log.Error(err, "Failed to update prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    ...
    ...
}
```

- Please review the [observability-best-practices] for additional important information regarding alerts and recording rules.

#### Alerts Unit Testing
It is highly recommended implementing unit tests for prometheus rules. For more information, please follow the Prometheus [unit testing documentation]. For examples of unit testing in a Golang operator, see the sample memcached-operator [alerts unit tests].

#### Inspecting the metrics, alerts and recording rules with Prometheus UI
Finally, in order to inspect the exposed metrics and alerts, we need to forward the corresponding port where metrics are published by Prometheus (usually `9090`, which is the default value). This can be done with the following command:
```bash
$ kubectl -n monitoring port-forward svc/prometheus-k8s 9090
```

where we assume that the prometheus service is available in the `monitoring` namespace.

> Now you can access Prometheus UI using [http://localhost:9090](http://localhost:9090).

Author: @everettraven, @varshaprasad96 I think sanity fails because this URL is not accessible. How can we handle that?

Author: @everettraven @varshaprasad96 any idea?

Reviewer 3: As a temporary workaround, we need not add it as a link.
```suggestion
Now you can access Prometheus UI. It is usually available at `http://localhost:9090`. For more details on exposing prometheus metrics, please [refer](https://prometheus.io/docs/prometheus/latest/getting_started/#starting-prometheus).
```

Author: Done

[PrometheusRule reconciliation example]:

```go
func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
    ...
    ...
    // Check if prometheus rule already exists, if not create a new one
    foundRule := &monitoringv1.PrometheusRule{}
    err := r.Get(ctx, types.NamespacedName{Name: ruleName, Namespace: namespace}, foundRule)
    if err != nil && apierrors.IsNotFound(err) {
        // Define a new prometheus rule
        prometheusRule := monitoring.NewPrometheusRule(namespace)
        if err := r.Create(ctx, prometheusRule); err != nil {
            log.Error(err, "Failed to create prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    
    if err == nil {
    // Check if prometheus rule spec was changed, if so set as desired
    desiredRuleSpec := monitoring.NewPrometheusRuleSpec()
    if !reflect.DeepEqual(foundRule.Spec.DeepCopy(), desiredRuleSpec) {
        desiredRuleSpec.DeepCopyInto(&foundRule.Spec)
        if r.Update(ctx, foundRule); err != nil {
            log.Error(err, "Failed to update prometheus rule")
            return ctrl.Result{}, nil
        }
    }
    ...
    ...
}
```

> - For more information regarding alerts and recording rules best practices, please follow [observability-best-practices].

Author: Done

> - It is highly recommended implementing unit tests for prometheus rules. For more information, please follow Prometheus [unit testing documentation]. For a reference of unit tests implementation in a Golang operator, see memcached-operator [alerts unit tests].

Reviewer: Please have this under its own header

Author: done
